all.equal no longer error when not a DT

d-sci opened the pull request:
Closes #4042 
Fix so that `all.equal(DT, not_a_DT)` returns the typical messages rather than error.
I had to modify old tests 1613.59 and 1613.60, which were checking for this error.

MichaelChirico commented on R/setops.R:
```diff
@@ -103,8 +103,8 @@ fsetequal = function(x, y, all=TRUE) {
-  stopifnot(is.logical(trim.levels), is.logical(check.attributes), is.logical(ignore.col.order), is.logical(ignore.row.order), is.numeric(tolerance))
-  if (!is.data.table(target) || !is.data.table(current)) stop("'target' and 'current' must both be data.tables")
+  stopifnot(is.logical(trim.levels), is.logical(check.attributes), is.logical(ignore.col.order), is.logical(ignore.row.order), is.numeric(tolerance), is.data.table(target))
+  if (!is.data.table(current)) return(NextMethod())
```
This will return `all.equal(data.table(a=1), data.frame(a=1))` to `TRUE` right? Shouldn't that be `FALSE` with reason "`current` is not a `data.table`"?

d-sci replied:
It will return
```
[1] "Attributes: < Names: 2 string mismatches >"                              
[2] "Attributes: < Length mismatch: comparison on first 2 components >"       
[3] "Attributes: < Component 1: target is externalptr, current is character >"
[4] "Attributes: < Component 2: Modes: character, numeric >"                  
[5] "Attributes: < Component 2: Lengths: 2, 1 >"                              
[6] "Attributes: < Component 2: target is character, current is numeric >"    
```

which is the same (just with some terms swapped) as the result of `all.equal(data.frame(a=1), data.table(a=1))`

jangorecki commented on R/setops.R:
```diff
@@ -103,8 +103,8 @@ fsetequal = function(x, y, all=TRUE) {
-  stopifnot(is.logical(trim.levels), is.logical(check.attributes), is.logical(ignore.col.order), is.logical(ignore.row.order), is.numeric(tolerance))
-  if (!is.data.table(target) || !is.data.table(current)) stop("'target' and 'current' must both be data.tables")
+  stopifnot(is.logical(trim.levels), is.logical(check.attributes), is.logical(ignore.col.order), is.logical(ignore.row.order), is.numeric(tolerance), is.data.table(target))
+  if (!is.data.table(current)) return(NextMethod())
```
why dispatch to next method instead of returning scalar character about class mismatch? if `target` is a data.table then it make sense to produce an answer in data.table method rather than redirecting to another.

d-sci replied:
I guess I thought that `all.equal(DT, DF)` should return roughly the same info as `all.equal(DF, DT)`. 

But happy to do the just do that if you prefer.

jangorecki replied:
I would rather prefer to return scalar early rather than investigating all differences, `all.equal` can be expensive, especially when you use `ignore.row.order`. Of course `check.attributes=FALSE` should continue to work, the same way as `all.equal(DF, DT, check.attributes=FALSE)` do.

d-sci commented:
@jangorecki raised an interesting point with regards to `check.attributes = F`. And what the result of `all.equal` should be on other-wise identical `data.table` vs `data.frame`. Base R can actually be oddly inconsistent here on `all.equal(x, y)` vs `all.equal(y,x)`:

```r
all.equal(data.frame(a=1), c(a=1), check.attributes = F)
#> [1] TRUE
all.equal(c(a=1), data.frame(a=1), check.attributes = F)
#> [1] "target is numeric, current is data.frame"
```

Current data.table always errors when the second argument isn't a data.table (regardless of `check.attributes`), and most of `all.equal.data.table` seems written based on the assumption that both args are data.tables. With new commit, this PR takes an easy middle ground, relaxing the error to a message, but still not returning TRUE for `all.equal(data.table(a=1), data.frame(a=1), check.attributes = F`), which would require more work:

```r
all.equal(data.table(a=1), c(a=1), check.attributes = F) # would ERROR before
#> [1] "target is data.table, current is numeric"
all.equal(c(a=1), data.table(a=1), check.attributes = F)
#> [1] "target is numeric, current is data.table"

all.equal(data.table(a=1), data.frame(a=1), check.attributes = F) # would ERROR before
#> [1] "target is data.table, current is data.frame"
all.equal(data.frame(a=1), data.table(a=1), check.attributes = F)
#> [1] TRUE
```

jangorecki commented on R/setops.R:
```diff
@@ -103,8 +103,8 @@ fsetequal = function(x, y, all=TRUE) {
-  stopifnot(is.logical(trim.levels), is.logical(check.attributes), is.logical(ignore.col.order), is.logical(ignore.row.order), is.numeric(tolerance))
-  if (!is.data.table(target) || !is.data.table(current)) stop("'target' and 'current' must both be data.tables")
+  stopifnot(is.logical(trim.levels), is.logical(check.attributes), is.logical(ignore.col.order), is.logical(ignore.row.order), is.numeric(tolerance), is.data.table(target))
+  if (!is.data.table(current)) return(paste0('target is data.table, current is ', data.class(current)))
```
what about `check.attributes=FALSE`?

d-sci replied:
`all.equal(data.frame(a=1), c(a=1), check.attributes = F)` returns TRUE -- do you think it should for `data.table` as well? 

The problem with continuing on despite `current` not being a data.table (could be as simple as a numeric, in that example) is that you have to a lot more careful about calling `setcolorder`, `nrow`, etc, on it which happens in the next few lines. Hence my original idea to just call `NextMethod`. I saw this as an easy middle ground. Maybe if `check.attributes = FALSE` just do `current <- as.data.table(current)` then keep going??

jangorecki replied:
class is an attribute. `NextMethod` could eventually be an option but for a `check.attributes=FALSE` then. Also we need to take care and check if any extra arguments haven't been used, because they will be ignored in a non-data.table all.equal method.

d-sci replied:
Ok, I'll try to think about this some more. As is, I do believe it is still an improvement, though clearly not perfect.

jangorecki replied:
It is improvement, but I don't think we should merge it like this without support for `check.attributes` argument.

d-sci commented:
Ok, new attempt to handle `check.attributes` appropriately. Basically relying on the fact that `setDT(x)` does a good job of checking whether `x`is at all data.table-like and if so, coercing to a data.table without actually changing any non-attribute data. So, if `current` is not a data.table:
- if `check.attributes = TRUE` return quickly with class mismatch info
- if `check.attributes = FALSE` but`current` is nothing like a data.table, return quickly with message to that effect
- if `check.attributes = FALSE` but`current` can be coerced via `setDT`, do so and continue on (thus still honouring `ignore.col.order`, `ignore.row.order` etc.)

jangorecki commented on R/setops.R:
```diff
@@ -103,8 +103,13 @@ fsetequal = function(x, y, all=TRUE) {
+  if (!is.data.table(current)) {
+    if (check.attributes) return(paste0('target is data.table, current is ', data.class(current)))
+    try({current = setDT(copy(current))}, silent = TRUE)
+    if (!is.data.table(current)) return('target is data.table and current is not of a comparable form')
```
Instead of `comparable form` I would use working that `current was not a data.table and it failed to be coerced to it`.
Also `as.data.table` seems to be more appropriate than `setDT(copy(.))`, as it may handle any class than will provide `as.data.table` method.

jangorecki commented on inst/tests/tests.Rraw:
```diff
@@ -8466,8 +8466,17 @@ test(1613.573, all.equal(cbind(x, factor(1)), cbind(x, factor(1)), ignore.row.or
+# fix for #4042
+test(1613.59, all.equal.data.table(1L, 2L), error = "is.data.table(target) is not TRUE")
+test(1613.601, all.equal(data.table(a=1), data.frame(a=1)), "target is data.table, current is data.frame")
+test(1613.602, all.equal(data.table(a=1), data.frame(a=1), check.attributes = F))
```
those `F` needs to be `FALSE`

jangorecki commented on inst/tests/tests.Rraw:
```diff
@@ -8466,8 +8466,17 @@ test(1613.573, all.equal(cbind(x, factor(1)), cbind(x, factor(1)), ignore.row.or
+test(1613.605, all.equal(data.table(a=1), 1L, check.attributes = F), "target is data.table and current is not of a comparable form")
+L1 <- list(a = data.table(1L), b = setattr("foo", "tbl", data.table(1L)))
+L2 <- list(a = list(1L), b = setattr("foo", "tbl", 1L))
+test(1613.606, is.character(all.equal(L1, L2)))
```
instead of just `is.character` maybe testing for a string, then tests will be also self describing itself.

d-sci commented:
Addressed most of your comments as suggested. I didn't change `copy(setDT(x))` to `as.data.table(x)` because `as.data.table` seems like it could do quite a lot of coercion that really changes underlying values. In particular, it means that both `list(a = 1, b = c(2,2))` and `list(a = c(1,1), b = c(2,2))` would be "all equal" with `data.table(a = c(1,1), b = c(2,2))` even though they wouldn't be with each other (since `as.data.table` can expand columns but `setDT` will error if any list elements are unequal length). `setDT` seems to be really good at telling "what could possibly be a data.table as it currently is, even if it' class isn't data.table", and I can't see a situation where it would fail on `X` but you would still expect `all.equal(DT, X)`

jangorecki commented:
@d-sci thanks, I am getting your point. The thing is that if we supply a list into `current` one might actually expect it to be expanded, because this is how the list behaves when being coerced to data.table. If one wants to have behavior you described then probably `all.equal(as.list(target), current)` is more appropriate (then of course we dispatch to `all.equal.list`).
If `as.data.table` way handles your use case then I would go with it. Otherwise probably best to hear what others think about `try()` coercion and the consistency of list/data.table equality.

d-sci commented:
Ok, fair enough, switched over to `as.data.table`. This will certainly work for my use case -- as long as I get a string and not an error for `all.equal(DT, not_a_DT)` I'm happy!

jangorecki commented:
We need documentation update for `current` argument, to explicitly mention `as.data.table` method used for coercion when `check.attributes=F`. Probably also an example of `all.equal(dt(a=1), 1, check.attributes=FALSE))` which might be surprising.

jangorecki commented:
Thanks for fixing missing S3 registration. Unless you have something more in mind I think we can remove WIP label already.

d-sci commented:
Yep, this is ready in my mind! Thanks for all your comments and suggestions.

mattdowle commented:
Thanks, @d-sci. I added you to DESCRIPTION in this PR and invited you to be project member. Please accept the invite displayed in your GitHub repositories page and then you can create branches in the main project so we can more easily commit to each other's branches. Welcome!
